### ISSUE TITLE
undistort distCoeffs: difference between doc & code

### ISSUE THREAD

**Issue by MironMA** [opencv/opencv#20030]
It has come to my attention that the [following term](https://github.com/opencv/opencv/blob/7de627c504a3b8aa16cdc3de56efdc358df4b061/modules/calib3d/src/undistort.dispatch.cpp#L378):

`double icdist = (1 + ((k[7]*r2 + k[6])*r2 + k[5])*r2)/(1 + ((k[4]*r2 + k[1])*r2 + k[0])*r2);`

that can be found inside some distortion functions' internal algorithm on multiple occasions might diverge from [the documentation](https://docs.opencv.org/4.2.0/d9/d0c/group__calib3d.html#ga55c716492470bfe86b0ee9bf3a1f0f7e) where one can read:

![image](https://user-images.githubusercontent.com/68554285/116961646-e51b8500-ac71-11eb-9a4d-6a63794a2c61.png)

`icdist` is later on multiplied (not divided) with `x`. That is, given `k`, this term seems inverted according to the doc. I was porting some manual code to open-cv and only when I followed the logic of the code order on `k` was open-cv's own function able to correct a test image with the same coefficients I had previously collected and used on the "manual" code.

I could not find this issue elsewhere in this repository and I hope I'm not wasting anyone's time here.

**Comment by alalek** [comment]
> icdist

`i` means `inverse` in that name I suppose.

`distortion` vs `undistortion`

`k[]` array in code has special meaning (`distCoeffs` parameter), see docs.

What is suggestion to change?

---

Usage questions should go to Users OpenCV Q/A forum: https://forum.opencv.org/

**Comment by MironMA** [comment]
Yes, I get that `i` stands for `inverse`, but it's name doesn't matter as much as the final logical result of the operation. ` cvUndistortPointsInternal`  should provide the same internal operation as the one mentionned in `undistort` and al and I don't think it does.

That is, all of these undistortion operations will work since the order of k has no effect whatsoever, but when one try to use these coefficients with some external code or use his own coefficients with opencv's code and trusting the doc, it won't work.

Now since many might have previously gather some coefficients from opencv's code, I suggest that we update the doc accordingly instead of the code. However, please not that at a glance, it seems that the original doc's version is more frequently used (I have no specific exemple at hand).

**Comment by MironMA** [comment]
I agree we must make sense of this code before marking it as an issue. If my remark is unfounded, there is nothing to be done, of course; While if it is, I would suggest to change the doc, not the code, as some codes might have been previously built on these coefficients.

For ease of understanding, let's rename the k's in the code "ck" because they are not the same as the "k's" in the doc, but instead equal to all distortion coefficients.

The correspondence between the "ck" in the code and all the coefficients in the doc should be as follows:

`ck0, ck1, [...], ck13 == k1,k2,p1,p2,k3,k4,k5,k6,s1,s2,s3,s4,τx,τy`

For simplicity, suppose `p1, p2` and `k4+` are of value `0`, but that `k1`, `k2` and `k3` (`ck0, ck1, ck4`) are provided and are positive. We should therefore have:

`ck0, ck1, [...], ck7 == k1,k2,0,0,k3,0,0,0`

Let us also assume that the function is used with `cv::TermCriteria::COUNT`, as in line 444. 

The first calculation on x is on line 358 with `x = (x - cx)*ifx` which is suitable for the first line of procedural operations in the doc. The same is true for the tilt distortion compensation on line 365.

Then, simplified for the first run (subsequent runs also gave the same value) :

```
double icdist = 1/(1 + ((ck[4]*r2 + ck[1])*r2 + ck[0])*r2);
deltaX = 0
x = x0*icdist = x0/(1 + ((ck[4]*r2 + ck[1])*r2 + ck[0])*r2);
```

Where 'x0 = x' in the doc. And then everything under the scope of the ` if` from line 390 is ignored. But at this point, according to the doc, we should instead have:

`x = x0*(1 + ((ck[4]*r2 + ck[1])*r2 + ck[0])*r2);`

To get the result that the doc prescribes, one would have to set:

`ck0, ck1, [...], ck7 == 0, 0, 0, 0, 0, k1, k2, k3`

Agree or not?

**Comment by MironMA** [comment]
( to follow on the doc side, better go [here](https://docs.opencv.org/4.2.0/d9/d0c/group__calib3d.html#ga7dfb72c9cf9780a347fbe3d1c47e5d5a) )
